pkg/sql/catalog/descpb/index.go

```diff
@@ -81,13 +81,29 @@ func (desc *IndexDescriptor) explicitColumnIDsWithoutShardColumn() ColumnIDs {
 	return colIDs
 }
 
+// keyColumnIDsWithoutShardColumn returns implicit and explicit column ids of
+// the index excluding the shard column.
+func (desc *IndexDescriptor) keyColumnIDsWithoutShardColumn() ColumnIDs {
+	if !desc.IsSharded() {
+		return desc.KeyColumnIDs
+	}
+	colIDs := make(ColumnIDs, 0, len(desc.KeyColumnIDs)-1)
+	for i := range desc.KeyColumnNames {
+		if desc.KeyColumnNames[i] != desc.Sharded.Name {
+			colIDs = append(colIDs, desc.KeyColumnIDs[i])
+		}
+	}
+	return colIDs
+}
+
 // IsValidReferencedUniqueConstraint  is part of the UniqueConstraint interface.
 // It returns whether the index can serve as a referenced index for a foreign
 // key constraint with the provided set of referencedColumnIDs.
 func (desc *IndexDescriptor) IsValidReferencedUniqueConstraint(referencedColIDs ColumnIDs) bool {
 	return desc.Unique &&
 		!desc.IsPartial() &&
-		desc.explicitColumnIDsWithoutShardColumn().PermutationOf(referencedColIDs)
+		(desc.explicitColumnIDsWithoutShardColumn().PermutationOf(referencedColIDs) ||
+			desc.keyColumnIDsWithoutShardColumn().PermutationOf(referencedColIDs))
```

Comment:
There seems to be some redundancy here, in that the columns IDs returned by `keyColumnIDsWithoutShardColumn` always form a superset of `explicitColumnIDsWithoutShardColumn`. It seems that what we really want to check is that the referenced column IDs minus the shard and implicit columns equals the key column IDs. This makes referencing any implicit columns possible but not required. Of course right now there's only at most one implicit column but this change seems worth making if only to clarify our intentions in the code.

I'm not sure why the shard column is treated specially but perhaps I'm unaware of something. I don't see why it should be special but I don't have a strong opinion on this matter.